File: bindings/python/examples/echo_example/src/echo_example.py
#!/usr/bin/env -P /usr/bin:/usr/local/bin python3 -B
# coding: utf-8

# =========================================================================
# echo_example.py
#
# Copyright (c) the Contributors as noted in the AUTHORS file.
# This file is part of Ingescape, see https://github.com/zeromq/ingescape.
# 
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
# =========================================================================


import ingescape as igs
import sys


class Singleton(type):
    _instances = {}
    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
        return cls._instances[cls]


class EchoExample(metaclass=Singleton):
    def __init__(self):
        # inputs
        self.boolI = None
        self.integerI = None
        self.doubleI = None
        self.stringI = None
        self.dataI = None

        # outputs
        self._boolO = True
        self._integerO = 0
        self._doubleO = 0
        self._stringO = ""
        self._dataO = b'\x00'

    # outputs
    @staticmethod
    def set_impulsionO(self):
        igs.output_set_impulsion("impulsion")

    @property
    def boolO(self):
        return self._boolO

    @boolO.setter
    def boolO(self, value):
        self._boolO = value
        if self._boolO is not None:
            igs.output_set_bool("bool", self._boolO)
    @property
    def integerO(self):
        return self._integerO

    @integerO.setter
    def integerO(self, value):
        self._integerO = value
        if self._integerO is not None:
            igs.output_set_int("integer", self._integerO)
    @property
    def doubleO(self):
        return self._doubleO

    @doubleO.setter
    def doubleO(self, value):
        self._doubleO = value
        if self._doubleO is not None:
            igs.output_set_double("double", self._doubleO)
    @property
    def stringO(self):
        return self._stringO

    @stringO.setter
    def stringO(self, value):
        self._stringO = value
        if self._stringO is not None:
            igs.output_set_string("string", self._stringO)
    @property
    def dataO(self):
        return self._dataO

    @dataO.setter
    def dataO(self, value):
        self._dataO = value
        if self._dataO is not None:
            igs.output_set_data("data", value)

    # services
    def receive_values(self, sender_agent_name, sender_agent_uuid, boolV, integer, double, string, data, token, my_data):
        igs.info(f"Service receive_values called by {sender_agent_name} ({sender_agent_uuid}) with argument_list {boolV, integer, double, string, data} and token '{token}''")

    def send_values(self, sender_agent_name, sender_agent_uuid, token, my_data):
        print(f"Service send_values called by {sender_agent_name} ({sender_agent_uuid}), token '{token}' sending values : {self.boolO, self.integerO, self.doubleO, self.stringO, self.dataO}")
        igs.info(sender_agent_uuid, "receive_values", (self.boolO, self.integerO, self.doubleO, self.stringO, self.dataO), token)

File: bindings/javascript-node/examples/echo_example/src/echoExample.js
/*  =========================================================================
    echoExample.js 

    Copyright (c) the Contributors as noted in the AUTHORS file.
    This file is part of Ingescape, see https://github.com/zeromq/ingescape.

    This Source Code Form is subject to the terms of the Mozilla Public
    License, v. 2.0. If a copy of the MPL was not distributed with this
    file, You can obtain one at http://mozilla.org/MPL/2.0/.
    =========================================================================
*/


const igs = require('ingescape');

class EchoExample {

  constructor() {
    //inputs
    this.boolI = false;
    this.integerI = 0;
    this.doubleI = 0;
    this.stringI = "";
    this.dataI = null;

    //outputs
    this.boolO = false;
    this.integerO = 0;
    this.doubleO = 0;
    this.stringO = "";
    this.dataO = null;
  }

  /////////////////////////////////////////////////////////////////////
  //inputs

  // impulsion
  setImpulsionI() {
    this.setImpulsionO();
  }

  // bool
  setBoolI(boolI) {
    this.boolI = boolI;
    this.setBoolO(this.boolI);
  }
  getBoolI() {
    return this.boolI;
  }

  // integer
  setIntegerI(integerI) {
    this.integerI = integerI;
    this.setIntegerO(this.integerI);
  }
  getIntegerI() {
    return this.integerI;
  }

  // double
  setDoubleI(doubleI) {
    this.doubleI = doubleI;
    this.setDoubleO(this.doubleI);
  }
  getDoubleI() {
    return this.doubleI;
  }

  // string
  setStringI(stringI) {
    this.stringI = stringI;
    this.setStringO(this.stringI);
  }
  getStringI() {
    return this.stringI;
  }

  // data
  setDataI(dataI) {
    this.dataI = dataI;
    this.setDataO(this.dataI);
  }
  getDataI() {
    return this.dataI;
  }

  /////////////////////////////////////////////////////////////////////
  //outputs

  // impulsion
  setImpulsionO() {
    igs.outputSetImpulsion("impulsion");
  }

  // bool
  setBoolO(boolO) {
    this.boolO = boolO;
    igs.outputSetBool("bool", this.boolO);
  }
  getBoolO() {
    return this.boolO;
  }

  // integer
  setIntegerO(integerO) {
    this.integerO = integerO;
    igs.outputSetInt("integer", this.integerO);
  }
  getIntegerO() {
    return this.integerO;
  }

  // double
  setDoubleO(doubleO) {
    this.doubleO = doubleO;
    igs.outputSetDouble("double", this.doubleO);
  }
  getDoubleO() {
    return this.doubleO;
  }

  // string
  setStringO(stringO) {
    this.stringO = stringO;
    igs.outputSetString("string", this.stringO);
  }
  getStringO() {
    return this.stringO;
  }

  // data
  setDataO(dataO) {
    this.dataO = dataO;
    igs.outputSetData("data", this.dataO);
  }
  getDataO() {
    return this.dataO;
  }

  /////////////////////////////////////////////////////////////////////
  //services

  // receive_values
  receiveValues(bool, integer, double, string, data) {
    igs.info("Receives values : ");

    igs.info("bool : " + bool);
    igs.info("integer : " + integer);
    igs.info("double : " + double);
    igs.info("string : " + string);
    igs.info("data with byte length : " + (data == null ? 0 : data.byteLength));
  }

  // send_values
  sendValues() {
    let argsList = [];
    argsList = igs.serviceArgsAddBool(argsList, this.boolI);
    argsList = igs.serviceArgsAddInt(argsList, this.integerI);
    argsList = igs.serviceArgsAddDouble(argsList, this.doubleI);
    argsList = igs.serviceArgsAddString(argsList, this.stringI);
    argsList = igs.serviceArgsAddData(argsList, this.dataI);
    igs.serviceCall(toAgent, "receive_values", argsList, "");
  }
};

module.exports = {EchoExample:EchoExample};

